### DISCUSSION TITLE
Rowing phase detection

### DISCUSSION THREAD

**Discussion by KjellVerb** [JaapvanEkris/openrowingmonitor#60] [Rower Profile Development]
Hi,

After my first full session with ORM, I got some unexpected results so I decided to dive into the data logs, where I saw unexpectedly long drive phases:
```
10/30/2024, 9:58:32 PM | npm[923] | stroke:  409, dist: 4016.7m, speed: 3.54m/s, pace: 2:21/500m, power: 124W, cal:  233.0kcal, SPM: 22.7, drive dur: 1.35s, rec. dur: 1.27s, stroke dur:  2.64s
10/30/2024, 9:58:33 PM | npm[923] | *** RECOVERY phase started at time: 1114.1413 sec
10/30/2024, 9:58:35 PM | npm[923] | *** DRIVE phase started at time: 1115.6762 sec
10/30/2024, 9:58:35 PM | npm[923] | *** Calculated drag factor: 279.3795, no. samples: 122, Goodness of Fit: 0.9885
10/30/2024, 9:58:35 PM | npm[923] | *** Calculated recovery slope: 0.005217, Goodness of Fit: 0.9885
10/30/2024, 9:58:35 PM | npm[923] | stroke:  410, dist: 4026.5m, speed: 3.53m/s, pace: 2:21/500m, power: 123W, cal:  233.6kcal, SPM: 22.6, drive dur: 1.34s, rec. dur: 1.31s, stroke dur:  2.66s
10/30/2024, 9:58:36 PM | npm[923] | *** RECOVERY phase started at time: 1117.0236 sec
10/30/2024, 9:58:38 PM | npm[923] | *** DRIVE phase started at time: 1118.5359 sec
10/30/2024, 9:58:38 PM | npm[923] | *** Calculated drag factor: 280.3024, no. samples: 120, Goodness of Fit: 0.9888
10/30/2024, 9:58:38 PM | npm[923] | *** Calculated recovery slope: 0.005234, Goodness of Fit: 0.9888
```

This prompted me to look around in the .csv's (Raw and RowingData). After some graphing around, I discovered that the auto-calculated drag factor drifts a lot and the driveTime variation is closely linked to that. This could also be the other way around of course because the dragfactor calculation has less of a recovery phase duration to work with if the drive phase is incorrectly detected too long.
I checked and `autoAdjustRecoverySlope` is set to `false` so I don't think the culprit is there. 

Does anyone have an idea of what to check next?

![afbeelding](https://github.com/user-attachments/assets/2c39a48c-bf99-4e96-bd0b-336772252e6a)

**Comment by Abasz** [comment]
Di you have the raw data file? I like to graph those in excel to see if they are consistent as well as it is possible to replay a session which helps tweaking the settings without needing to row in real time 

It is possible that you need stricter goodness of fit for the drag factor calculation.

**Reply by KjellVerb** [reply]
[2024-10-30_20-39-59_raw.csv.gz](https://github.com/user-attachments/files/17590383/2024-10-30_20-39-59_raw.csv.gz)
[2024-10-30_20-39-59_rowingData.csv](https://github.com/user-attachments/files/17590384/2024-10-30_20-39-59_rowingData.csv)
[2024-10-30_20-39-59_Analysis.xlsx](https://github.com/user-attachments/files/17590434/2024-10-30_20-39-59_Analysis.xlsx)


Luckily I have both. I've already tried plotting the stroke detection vs the raw data but it doesn't show anything obvious.
How does replaying work? I found some instructions in readme but they were quite brief and I didn't find how to apply it.

**Reply by Abasz** [reply]
replay: in server.js at the end you will find the replay function. Its commented out you need to uncomment.

```js
replayRowingSession(handleRotationImpulse, {
  filename: 'recordings/Concept2_RowErg_Session_2000meters.csv', // Example row from a Concept 2 RowErg, 2000 meters
  realtime: true,
  loop: false
})
```

Set the path the raw data csv file. Need ot ungz of course

Set realtime to false.

run npm start (I recommend closing the GUI if you have it open

It will replay the session as fast as it can (it will generated csvfiles like it would in normal mode). 

Then tweak settings and check again. Rinse and repeat until you are good to go.

**Reply by KjellVerb** [reply]
Just as you commented I had it figured out, thanks for the comment and sorry for the wasted effort.
Looks like I THOUGHT I had disabled `autoAdjustRecoverySlope` but maybe I didn't save config.js before restarting the service because the log already is different.
OLD: `10/30/2024, 9:58:35 PM | npm[923] | *** Calculated recovery slope: 0.005217, Goodness of Fit: 0.9885`
NEW: `10/31/2024, 10:21:16 PM | npm[3319] | *** Calculated recovery slope: 0.004882, Goodness of Fit: 0.9736, not used as autoAdjustRecoverySlope isn't set to true`

I did not plot anything in excel yet but the driveTimes in the logs are all very consistently between 0.68 and 0.79 so looks like my original issue is solved.

This means that for my current settings `autoAdjustRecoverySlope` does not work or there is some sort of feedback loop between the recovery slope calculation, the stroke phase detection, and the drag factor calculation.

**Reply by JaapvanEkris** [reply]
> This means that for my current settings `autoAdjustRecoverySlope` does not work or there is some sort of feedback loop between the recovery slope calculation, the stroke phase detection, and the drag factor calculation.

There is, by definition. The dragfactor and recoveryslope (which in essence are the same thing) are calculated during the recovery. When the stroke detection has issues, it will carry over into the recovery, which isn't recognized optimally, resulting in a badly calculated dragfactor/recoveryslope.

We combat this issue in several ways, first by using an algorithm with a known noise resistance, second by having that same algorithm provide a quality indicator (aka Goodness of Fit) which is used to filter these calculated values, and third by using a running weighted average across several strokes. This should stabilize the calculated dragfactor etc..

However, water based rowers and similar machines have odd drag behaviour, making these advanced approaches unsuitable. Thus, you have to revert to the acceleration/deceleration approach, which is robust but less precise (please note, a Concept2 PM5 uses this less precise approach as well).

**Reply by KjellVerb** [reply]
The acc/dec approach works well indeed for stroke detection.
My machine has a magnetic brake so the drag force should be quite linear to angular velocity. My raw data (see post below) also looks really clean to me so I'm not sure where the algorithm would go wrong.
Even on a row without autoAdjustRecoverySlope (simulated on the same data) there is quite a big variation in dragfactor which seems to inversely impact the calculated power as well. The logs show a goodness of fit usually above 0.97 and always above 0.963
Any idea what I could improve here?
![afbeelding](https://github.com/user-attachments/assets/5854ac0f-6d33-48e0-97a0-c4ac8756e5e5)

**Reply by Abasz** [reply]
Magnetic break force is Indeed linear. Which means on slower rotation speeds the applied drag factor is higher than on higher speed.

Thus also means that when you do a stronger pull your recovery slope will shallower, while slower stroke steeper. This is why autoadjust does not work and DF moves around. As if there is too much moving around it is not able to detect (you simply never reach the set minimum recovery slope value 

I experienced the same issue with old flywheels where there is a lot of bearing and mechanical drag

**Reply by JaapvanEkris** [reply]
> The acc/dec approach works well indeed for stroke detection. My machine has a magnetic brake so the drag force should be quite linear to angular velocity. My raw data (see post below) also looks really clean to me so I'm not sure where the algorithm would go wrong. Even on a row without autoAdjustRecoverySlope (simulated on the same data) there is quite a big variation in dragfactor which seems to inversely impact the calculated power as well.

The issue is that most machines are air/waterbracked, where the dragforce is a function of the angular velocity. On a magnetic rower, it is a constant force. So you need a different formula to make df stable, but as the df * velocity underpins many formulea, you essentially end up rewriting the entire engine.

Using a longer averaging for dragfactor, or a fixed value for df typically leads to satisfactory results.

**Reply by KjellVerb** [reply]
So in a way this should be quite easy to calculate from the raw data but magnetic rowers would require a separate physics model where the dragForce is autocalculated instead of the dragFactor?

**Reply by JaapvanEkris** [reply]
> So in a way this should be quite easy to calculate from the raw data but magnetic rowers would require a separate physics model where the dragForce is autocalculated instead of the dragFactor?

I wouldn't say easy, it is just different than we do, and AFAIK, nobody did this before. Removing the factor df * angular_velocity for each formula sounds easy, until the formula you are trying to adapt contains just one of the two elements, and not the other.

**Reply by KjellVerb** [reply]
Guess I'll just accept the inaccuracies (as they are consistent between sessions) and save up for a 2ndhand C2

**Comment by KjellVerb** [comment]
**Correctly** disabling `autoAdjustRecoverySlope` solved the issue, see here the before and after of the original RowingData and the re-simulated RowingData

![afbeelding](https://github.com/user-attachments/assets/7587b4c3-a424-4405-a51e-1e07af98fd3d)

![afbeelding](https://github.com/user-attachments/assets/6654f71c-8ed2-4584-addf-4a58f0a15678)
